File: apps/android/app/src/main/java/com/sanvya/app/ui/search/SearchViewModel.kt
package com.sanvya.app.ui.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.repository.Account
import com.sanvya.app.data.repository.CategoryRow
import com.sanvya.app.data.repository.LedgerRepository
import com.sanvya.app.data.repository.TransactionRow
import com.sanvya.app.domain.search.SearchCriteria
import com.sanvya.app.domain.search.SearchRow
import com.sanvya.app.domain.search.searchTransactions
import com.sanvya.app.domain.splits.SplitInfo
import com.sanvya.app.domain.splits.collapseSplitRowIds
import com.sanvya.app.domain.splits.splitInfoByTransaction
import com.sanvya.app.ui.transactions.TransactionListItem
import com.sanvya.app.ui.transactions.transactionListItem
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/** What the screen shows once the filter has run. */
data class SearchState(
    val items: List<TransactionListItem> = emptyList(),
    /**
     * Web counts the FILTERED rows, before collapse -- so two postings of one
     * split expense count as two here and render as one row below. Copied
     * deliberately: the count answers "how much matched", not "how many rows".
     */
    val resultCount: Int = 0,
    val accounts: List<Account> = emptyList(),
)

/**
 * Search -- ported from apps/web/app/search/page.tsx.
 *
 * The filter itself is `domain.search.searchTransactions`, vector-tested
 * against a reference implementation of web's `useMemo`. This view model's
 * whole job is to keep the six live queries that feed it and to turn the
 * surviving rows into the SAME [TransactionListItem] the Transactions list and
 * the dashboard tile render -- web renders the same `<TransactionTile>` on all
 * three, and a third row builder here is the re-inlining the component
 * inventory exists to prevent.
 *
 * Mirrors apps/ios/App/ViewModels/SearchViewModel.swift.
 */
class SearchViewModel : ViewModel(), KoinComponent {
    private val ledgerRepository: LedgerRepository by inject()

    private val _criteria = MutableStateFlow(SearchCriteria())
    val criteria: StateFlow<SearchCriteria> = _criteria.asStateFlow()

    private val _showFilters = MutableStateFlow(false)
    val showFilters: StateFlow<Boolean> = _showFilters.asStateFlow()

    fun setQuery(v: String) { _criteria.value = _criteria.value.copy(query = v) }
    fun setType(v: String) { _criteria.value = _criteria.value.copy(type = v) }
    fun setAccountId(v: String) { _criteria.value = _criteria.value.copy(accountId = v) }
    fun setFrom(v: String) { _criteria.value = _criteria.value.copy(from = v) }
    fun setTo(v: String) { _criteria.value = _criteria.value.copy(to = v) }
    fun setMin(v: String) { _criteria.value = _criteria.value.copy(min = v) }
    fun setMax(v: String) { _criteria.value = _criteria.value.copy(max = v) }
    fun toggleFilters() { _showFilters.value = !_showFilters.value }

    /** Clears every filter but keeps what was typed -- web's `clearFilters`. */
    fun clearFilters() { _criteria.value = SearchCriteria(query = _criteria.value.query) }

    private data class Source(
        val txns: List<TransactionRow>,
        val accounts: List<Account>,
        val categories: List<CategoryRow>,
        val labelNames: Map<String, List<String>>,
        val splitInfo: Map<String, SplitInfo>,
    )

    private val source: StateFlow<Source> = combine(
        ledgerRepository.watchSearchTransactions(),
        ledgerRepository.watchAccounts(includeArchived = true),
        ledgerRepository.watchCategories(),
        ledgerRepository.watchTransactionLabelNames(),
        ledgerRepository.watchSplitPostings(),
    ) { txns, accounts, categories, labelNames, postings ->
        Source(txns, accounts, categories, labelNames, splitInfoByTransaction(postings))
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = Source(emptyList(), emptyList(), emptyList(), emptyMap(), emptyMap()),
    )

    /**
     * The repository returns one row per (account type, method) pairing, so the
     * same method arrives several times. Only the id -> label mapping is wanted
     * here, and it is the same in every pairing.
     */
    private val methodLabels: StateFlow<Map<String, String>> =
        ledgerRepository.watchPaymentMethods()
            .map { methods -> methods.associate { it.id to it.label } }
            .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyMap())

    val state: StateFlow<SearchState> = combine(
        source, methodLabels, _criteria,
    ) { (txns, accounts, categories, labelNames, splitInfo), methods, criteria ->
        val accountMap = accounts.associateBy { it.id }
        val categoryMap = categories.associateBy { it.id }

        val rows = txns.map { txn ->
            val account = accountMap[txn.accountId]
            SearchRow(
                id = txn.id,
                type = txn.type,
                accountId = txn.accountId,
                toAccountId = txn.toAccountId,
                occurredAt = txn.occurredAt,
                amountMinor = txn.amount,
                currency = txn.currency,
                labels = labelNames[txn.id]?.joinToString(", "),
                note = txn.note,
                description = txn.description,
                methodLabel = txn.paymentMethod?.let { methods[it] },
                categoryName = txn.categoryId?.let { categoryMap[it]?.name },
                accountName = account?.name,
                accountType = account?.type,
            )
        }

        val matched = searchTransactions(rows, criteria)
        val byId = txns.associateBy { it.id }
        SearchState(
            items = collapseSplitRowIds(matched.map { it.id }, splitInfo).mapNotNull { id ->
                byId[id]?.let { txn ->
                    transactionListItem(txn, accountMap, categoryMap, labelNames[id], splitInfo[id])
                }
            },
            resultCount = matched.size,
            accounts = accounts,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), SearchState())
}

File: apps/ios/App/ViewModels/SearchViewModel.swift
import Foundation
import Observation
import Factory
import Data
import Domain

/// Search — ported from apps/web/app/search/page.tsx.
///
/// The filter itself is `Domain.searchTransactions`, vector-tested against a
/// reference implementation of web's `useMemo`. This view model's whole job is
/// to keep the six live queries that feed it and to turn the surviving rows
/// into the SAME `TransactionListItem` the Transactions list and the dashboard
/// tile render — web renders the same `<TransactionTile>` on all three, and a
/// third row builder here is the re-inlining the component inventory exists to
/// prevent.
///
/// Mirrors apps/android/.../ui/search/SearchViewModel.kt.
@Observable
@MainActor
final class SearchViewModel {
    @ObservationIgnored
    @Injected(\.ledgerRepository) private var ledgerRepository

    /// Mutating any field re-runs the filter — `didSet` fires for an in-place
    /// mutation of a struct property, which is what `criteria.query = "…"` is.
    var criteria = SearchCriteria() { didSet { recompute() } }
    var showFilters = false

    /// The account picker's options, in the order the repository returns them.
    private(set) var accounts: [Account] = []
    private(set) var items: [TransactionListItem] = []
    /// Web counts the FILTERED rows, before collapse — so two postings of one
    /// split expense count as two here and render as one row below. Copied
    /// deliberately: the count answers "how much matched", not "how many rows".
    private(set) var resultCount = 0

    private var tasks: [Task<Void, Never>] = []
    private var allTxns: [TransactionRow] = []
    private var accountMap: [String: Account] = [:]
    private var categoryMap: [String: CategoryRow] = [:]
    private var labelNames: [String: [String]] = [:]
    private var methodLabels: [String: String] = [:]
    private var splitInfo: [String: SplitInfo] = [:]

    var activeFilters: Int { activeFilterCount(criteria) }

    func clearFilters() {
        criteria = SearchCriteria(query: criteria.query)
    }

    /**
     Apply a deep link's filters — once.

     Web guards its prefill effect with a `prefilled` flag for a reason: the
     effect re-runs whenever `params` changes identity, and re-applying would
     wipe whatever the user had typed since arriving. The flag lives here rather
     than in the view because a view model outlives the redraws a `@State` in a
     recreated view does not.
     */
    func applyPrefill(_ prefill: SearchPrefill) {
        guard !prefilled else { return }
        prefilled = true
        criteria = prefill.criteria
        if prefill.showFilters { showFilters = true }
    }

    private var prefilled = false

    func start() {
        cancel()
        tasks.append(Task {
            do {
                for try await txns in try ledgerRepository.watchSearchTransactions() {
                    self.allTxns = txns
                    self.recompute()
                }
            } catch { print("Error watching search transactions: \(error)") }
        })
        tasks.append(Task {
            do {
                for try await accounts in try ledgerRepository.watchAccounts(includeArchived: true) {
                    self.accounts = accounts
                    self.accountMap = Dictionary(uniqueKeysWithValues: accounts.map { ($0.id, $0) })
                    self.recompute()
                }
            } catch { print("Error watching accounts: \(error)") }
        })
        tasks.append(Task {
            do {
                for try await cats in try ledgerRepository.watchCategories() {
                    self.categoryMap = Dictionary(uniqueKeysWithValues: cats.map { ($0.id, $0) })
                    self.recompute()
                }
            } catch { print("Error watching categories: \(error)") }
        })
        tasks.append(Task {
            do {
                for try await names in try ledgerRepository.watchTransactionLabelNames() {
                    self.labelNames = names
                    self.recompute()
                }
            } catch { print("Error watching transaction labels: \(error)") }
        })
        tasks.append(Task {
            do {
                // The repository returns one row per (account type, method)
                // pairing, so the same method arrives several times. Only the
                // id → label mapping is wanted here, and it is the same in
                // every pairing.
                for try await methods in try ledgerRepository.watchPaymentMethods() {
                    self.methodLabels = Dictionary(methods.map { ($0.id, $0.label) }) { first, _ in first }
                    self.recompute()
                }
            } catch { print("Error watching payment methods: \(error)") }
        })
        tasks.append(Task {
            do {
                for try await postings in try ledgerRepository.watchSplitPostings() {
                    self.splitInfo = splitInfoByTransaction(postings)
                    self.recompute()
                }
            } catch { print("Error watching split postings: \(error)") }
        })
    }

    func cancel() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private func recompute() {
        let rows = allTxns.map { txn -> SearchRow in
            let account = accountMap[txn.accountId]
            return SearchRow(
                id: txn.id,
                type: txn.type,
                accountId: txn.accountId,
                toAccountId: txn.toAccountId,
                occurredAt: txn.occurredAt,
                amountMinor: txn.amount,
                currency: txn.currency,
                labels: labelNames[txn.id]?.joined(separator: ", "),
                note: txn.note,
                description: txn.description,
                methodLabel: txn.paymentMethod.flatMap { methodLabels[$0] },
                categoryName: txn.categoryId.flatMap { categoryMap[$0]?.name },
                accountName: account?.name,
                accountType: account?.type
            )
        }

        let matched = searchTransactions(rows, criteria)
        resultCount = matched.count

        let byId = Dictionary(uniqueKeysWithValues: allTxns.map { ($0.id, $0) })
        items = collapseSplitRowIds(matched.map(\.id), splitInfo).compactMap { id in
            guard let txn = byId[id] else { return nil }
            return transactionListItem(
                txn,
                accountMap: accountMap,
                categoryMap: categoryMap,
                labels: labelNames[id],
                split: splitInfo[id]
            )
        }
    }
}
